implementations/rust/ockam/ockam/src/entity/profile.rs

```diff
@@ -31,40 +69,258 @@ impl Profile {
         identifier: ProfileIdentifier,
         change_history: ProfileChangeHistory,
         verification_policies: Vec<ProfileVerificationPolicy>,
+        vault: Arc<Mutex<dyn ProfileVault>>,
     ) -> Self {
         Profile {
             identifier,
             change_history,
             verification_policies,
+            vault,
         }
     }
 }
 
 impl Profile {
-    pub fn apply(&mut self, change_event: ProfileChangeEvent) {
-        if !self.verify(&change_event) {
-            return; // TODO: Throw error
+    pub fn create_with_key(
+        label: String,
+        key_type: ProfileKeyType,
+        key_purpose: ProfileKeyPurpose,
+        attributes: Option<ProfileEventAttributes>,
+        vault: Arc<Mutex<dyn ProfileVault>>,
+    ) -> Result<Self, ockam_core::Error> {
+        let attributes = attributes.unwrap_or(ProfileEventAttributes::new());
+
+        let mut v = vault.lock().unwrap();
+
+        // TODO: Should be customisable
+        let secret_attributes = SecretAttributes {
+            stype: SecretType::Curve25519,
+            persistence: SecretPersistence::Persistent,
+            length: 0,
+        };
+
+        let private_key = v.secret_generate(secret_attributes)?;
+        let public_key = v.secret_public_key_get(&private_key)?;
+
+        let change =
+            CreateKeyChange::new(label, key_type, key_purpose, public_key.as_ref().to_vec());
+        let prev_id = v.sha256(&[])?;
```

Comment:
I think we should hash the public key vs an empty byte array, otherwise, it will be the same value for all profiles.

Reply:
This is constant that indicates that you're looking at the first event, it's the only purpose for it and I think it's fine it's the same for everybody. I could have removed it, but that will convert previous_id to Option<> which I wanted to avoid.

Reply:
Okay, let's add a domain separation parameter that we hardcode for this so it's not just the hash of empty. It's good practice to have.

Reply:
agree